xilem_web/src/concurrent/interval.rs

```diff
@@ -0,0 +1,137 @@
+pub struct Interval<Callback, State, Action> {
+    ms: i32,
+    callback: Callback,
+    phantom: PhantomData<fn() -> (State, Action)>,
+}
+
+/// Start an interval which invokes `callback` every `ms` milliseconds
+///
+/// `ms` is of type `i32` because of the weird choice for it in [`setInterval`](https://developer.mozilla.org/en-US/docs/Web/API/setInterval#sect2)
+///
+/// # Examples
+///
+/// ```
+/// use xilem_web::{core::fork, concurrent::interval, elements::html::div, interfaces::Element};
+///
+/// fn timer(seconds: &mut u32) -> impl Element<u32> {
+///     fork(
+///         div(format!("{seconds} seconds have passed, since creating this view")),
+///         interval(
+///             1000, // in ms, when this changes, the interval is reset
+///             |seconds: &mut u32| *seconds += 1,
+///         )
+///     )
+/// }
+/// ```
+pub fn interval<State, Action, OA, Callback>(
+    ms: i32,
+    callback: Callback,
+) -> Interval<Callback, State, Action>
+where
+    State: 'static,
+    Action: 'static,
+    OA: OptionalAction<Action> + 'static,
+    Callback: Fn(&mut State) -> OA + 'static,
+{
+    Interval {
+        ms,
+        callback,
+        phantom: PhantomData,
+    }
+}
+
+pub struct IntervalState {
+    // Closures are retained so they can be called by environment
+    interval_fn: Closure<dyn FnMut()>,
+    interval_handle: i32,
+}
+
+fn start_interval(callback: &Closure<dyn FnMut()>, ms: i32) -> i32 {
+    web_sys::window()
+        .unwrap_throw()
+        .set_interval_with_callback_and_timeout_and_arguments_0(
+            callback.as_ref().unchecked_ref(),
+            ms,
+        )
+        .unwrap_throw()
+}
+
+fn clear_interval(handle: i32) {
+    web_sys::window()
+        .unwrap_throw()
+        .clear_interval_with_handle(handle);
+}
+
+impl<Callback, State, Action> ViewMarker for Interval<Callback, State, Action> {}
+
+impl<State, Action, Callback, OA> View<State, Action, ViewCtx, DynMessage>
+    for Interval<Callback, State, Action>
+where
+    State: 'static,
+    Action: 'static,
+    OA: OptionalAction<Action> + 'static,
+    Callback: Fn(&mut State) -> OA + 'static,
+{
+    type Element = NoElement;
+
+    type ViewState = IntervalState;
+
+    fn build(&self, ctx: &mut ViewCtx) -> (Self::Element, Self::ViewState) {
+        let thunk = ctx.message_thunk();
+        let interval_fn = Closure::new(move || thunk.push_message(()));
+        let state = IntervalState {
+            interval_handle: start_interval(&interval_fn, self.ms),
+            interval_fn,
+        };
+
+        (NoElement, state)
+    }
+
+    fn rebuild<'el>(
+        &self,
+        prev: &Self,
+        view_state: &mut Self::ViewState,
+        _: &mut ViewCtx,
+        (): Mut<'el, Self::Element>,
+    ) -> Mut<'el, Self::Element> {
+        if prev.ms != self.ms {
+            clear_interval(view_state.interval_handle);
+            view_state.interval_handle = start_interval(&view_state.interval_fn, self.ms);
```

Comment:
There's a discussion here about what the expected behaviour is. Whether you should start a new task from fresh, or act as-if the previous task had been started with the new time.

Reply:
You mean you initiated the discussion or is there really somewhere a discussion somewhere?^^ Or do you mean the stuff in `memoized_await`?

Reply:
I don't think it's trivial though, might require a `setTimeout` with the remaining time + the change of the interval I guess...

Reply:
So, as in this is a common discussion with this kind of API. I don't think we've had it within Linebender.

We need to document it carefully

Reply:
We may extend that behavior in the future, because I see reasons for both behaviors, I chose this because it was more simple, but I agree, I'll add this to the doc-comment, so that it's clear